### PR TITLE
Feat/sync with upstream

### PR DESCRIPTION
**Pull request by Gumbees**
Don't review yet. I'm comparing.

### CODE REVIEW

**Comment by Gumbees** [comment]
Confirmed functional with zero on 192_7.

File: crypto/equihash.h [hunk 1]
```diff
@@ -224,14 +228,16 @@ inline bool EhBasicSolve(unsigned int n, unsigned int k, const eh_HashState& bas
...
         return Eh96_3.BasicSolve(base_state, validBlock, cancelled);
-    } else if (n == 144 && k == 5) {
-        return Eh144_5.BasicSolve(base_state, validBlock, cancelled);
     } else if (n == 200 && k == 9) {
         return Eh200_9.BasicSolve(base_state, validBlock, cancelled);
     } else if (n == 96 && k == 5) {
         return Eh96_5.BasicSolve(base_state, validBlock, cancelled);
     } else if (n == 48 && k == 5) {
         return Eh48_5.BasicSolve(base_state, validBlock, cancelled);
+    } else if (n == 144 && k == 5) {
```
Comment (egyptianbman) [inline crypto/equihash.h]: Super minor micro optimization but can you shift these two to the top -- below 200,9 above is fine. There's no sense in checking for 96,5 and 48,5 on every solution when it's highly unlikely they'll ever be used.

File: crypto/equihash.h [hunk 2]
```diff
@@ -250,14 +256,16 @@ inline bool EhOptimisedSolve(unsigned int n, unsigned int k, const eh_HashState&
...
         return Eh96_3.OptimisedSolve(base_state, validBlock, cancelled);
-    } else if (n == 144 && k == 5) {
-        return Eh144_5.OptimisedSolve(base_state, validBlock, cancelled);
     } else if (n == 200 && k == 9) {
         return Eh200_9.OptimisedSolve(base_state, validBlock, cancelled);
     } else if (n == 96 && k == 5) {
         return Eh96_5.OptimisedSolve(base_state, validBlock, cancelled);
     } else if (n == 48 && k == 5) {
         return Eh48_5.OptimisedSolve(base_state, validBlock, cancelled);
+    } else if (n == 144 && k == 5) {
```
Comment (egyptianbman) [inline crypto/equihash.h]: Same here. Can probably shift 96,3 down.

File: crypto/equihash.h [hunk 3]
```diff
@@ -274,14 +282,16 @@ inline bool EhOptimisedSolveUncancellable(unsigned int n, unsigned int k, const
...
         ret = Eh96_3.IsValidSolution(base_state, soln);  \
-    } else if (n == 144 && k == 5) {                     \
-        ret = Eh144_5.IsValidSolution(base_state, soln); \
     } else if (n == 200 && k == 9) {                     \
         ret = Eh200_9.IsValidSolution(base_state, soln); \
     } else if (n == 96 && k == 5) {                      \
         ret = Eh96_5.IsValidSolution(base_state, soln);  \
     } else if (n == 48 && k == 5) {                      \
         ret = Eh48_5.IsValidSolution(base_state, soln);  \
+    } else if (n == 144 && k == 5) {                     \
```
Comment (egyptianbman) [inline crypto/equihash.h]: And here

**Review by egyptianbman — CHANGES_REQUESTED** [review]
Can you also run the tests to make sure they still pass? https://github.com/z-nomp/equihashverify/blob/master/package.json#L8

There's also https://github.com/z-nomp/equihashverify/blob/master/test.js you can run to test.

**Comment by Gumbees** [comment]
```
  equihashverify
Invalid solution: invalid collision length between StepRows
X[i]   = 006d480d49200247f30f7ac908fdcc07f31d0553bd0c24a30f7a150b1bbc
X[i+1] = 08b9270a630a00172c0999c30ec18003dda60cfce90a444801df3f03d3aa
    1) should return true for correct solution test
Invalid solution: invalid collision length between StepRows
X[i]   = 02e9aa05f54e0f9cb903a6ea0af4d9064bad00cd6603cb1103c48a0767dd
X[i+1] = 08b9270a630a00172c0999c30ec18003dda60cfce90a444801df3f03d3aa
    ✓ should return false for wrong solution test
Invalid solution: duplicate indices
    ✓ should return false for fake solution test
Invalid solution: duplicate indices
Invalid solution: invalid collision length between StepRows
X[i]   = 0f235709ed4c04f8a90a9a4300c9470b1541084d7303054f0b27a5023d3f
X[i+1] = 0b0eb9052be307966605040e01aaf005f44b0bf88a02dfec00342b0005d7
Invalid solution: invalid collision length between StepRows
X[i]   = 0b0eb9052be307966605040e01aaf005f44b0bf88a02dfec00342b0005d7
X[i+1] = 01c7a202c8450b6ab6013ca804dcc6082c960d32850a6acf0585ea04c8ae
Invalid solution: invalid collision length between StepRows
X[i]   = 0b3bde0309610edd7a0ab261018a060e989008925d0406c408ed480730e6
X[i+1] = 01b7c20676a500bc2b029a2f02a9f10ef7bc008784046a350de33d0ba996
Invalid solution: invalid collision length between StepRows
X[i]   = 01c7a202c8450b6ab6013ca804dcc6082c960d32850a6acf0585ea04c8ae
X[i+1] = 0620b40e4f44093b0e04dbc60a1f0d0f698f004fc10df78e0e4c8d0275d9
Invalid solution: invalid collision length between StepRows
X[i]   = 0962400b37bd04c093049f5902c1e906ef5c01dabd0d84f30634f30cb4b3
X[i+1] = 00a35304f44f0771a40e4b110211d9056e8c07c8840e25e10cfac000325d
Invalid solution: invalid collision length between StepRows
X[i]   = 01b7c20676a500bc2b029a2f02a9f10ef7bc008784046a350de33d0ba996
X[i+1] = 0626690c537f0bdf7905288600104001438d0a2a330be5db0e8ff10682ed
Invalid solution: invalid collision length between StepRows
X[i]   = 00baa00ad67902f8770960ef080f1b0a13a409a49d017c7b0fdf34089d10
X[i+1] = 03294903af4502a53a0a42d8058c130f0a610010e90836ea0659af07ab5a
Invalid solution: invalid collision length between StepRows
X[i]   = 0620b40e4f44093b0e04dbc60a1f0d0f698f004fc10df78e0e4c8d0275d9
X[i+1] = 0f235709ed4c04f8a90a9a4300c9470b1541084d7303054f0b27a5023d3f
Invalid solution: invalid collision length between StepRows
X[i]   = 032d2f025f7809f5020dda800df1670b478501ca120049a000e51b02e827
X[i+1] = 0b3bde0309610edd7a0ab261018a060e989008925d0406c408ed480730e6
    ✓ should return false for same numeric values tests
Invalid solution: invalid collision length between StepRows
X[i]   = 00a35304f44f0771a40e4b110211d9056e8c07c8840e25e10cfac000325d
X[i+1] = 0962400b37bd04c093049f5902c1e906ef5c01dabd0d84f30634f30cb4b3
Invalid solution: invalid collision length between StepRows
X[i]   = 05de480f66170e0ec508c3c2079d3d0341ac06e9490762aa0d44820097c4
X[i+1] = 00baa00ad67902f8770960ef080f1b0a13a409a49d017c7b0fdf34089d10
Invalid solution: invalid collision length between StepRows
X[i]   = 0626690c537f0bdf7905288600104001438d0a2a330be5db0e8ff10682ed
X[i+1] = 032d2f025f7809f5020dda800df1670b478501ca120049a000e51b02e827
Invalid solution: invalid collision length between StepRows
X[i]   = 0ff1110a49de04059d08e4d7080d3a034ce60828d8071e4a0c481d0c0215
X[i+1] = 05de480f66170e0ec508c3c2079d3d0341ac06e9490762aa0d44820097c4
Invalid solution: invalid collision length between StepRows
X[i]   = 03294903af4502a53a0a42d8058c130f0a610010e90836ea0659af07ab5a
X[i+1] = 0ff1110a49de04059d08e4d7080d3a034ce60828d8071e4a0c481d0c0215
Invalid solution: duplicate indices
    ✓ should return false for same alpha values test
Invalid solution length: 1344 (expected 100)
Invalid solution length: 1344 (expected 100)
Invalid solution length: 1344 (expected 100)
Invalid solution length: 1344 (expected 100)
Invalid solution length: 1344 (expected 100)
Invalid solution length: 1344 (expected 100)
    ✓ should return false for same alpha values test


  5 passing (53ms)
  1 failing

  1) equihashverify
       should return true for correct solution test:

      AssertionError [ERR_ASSERTION]: false == true
      + expected - actual

      -false
      +true

      at Context.<anonymous> (test.js:11:16)



```

Getting the 1 failure on previous commits as well before this pull request. All others pass. I'd like to patch the test to get all passing if the test is written wrong. @egyptianbman

**Comment by Gumbees** [comment]
I recommitted the reorganization because I messed something up the first time around causing collisions. It is fixed now.

**Review by egyptianbman — APPROVED** [review]
Yay, tests passing is always good! Thanks!

I'd love to add at least one success test for each param group but that is for another day.

**Pull request merged**